scipy/special/ellint_carlson_cpp_lite/ellint_arith_aux.hh

```diff
@@ -0,0 +1,195 @@
+#ifndef ELLINT_ARITH_AUX_HH_INCLUDED
+#define ELLINT_ARITH_AUX_HH_INCLUDED
+
+
+#include <cstddef>
+#include "ellint_typing.hh"
+
+
+/* Auxiliary floating-point manipulation utilities.
+ * Ref:
+ *
+ * S. M. Rump, T. Ogita, S. Oishi: Accurate Floating-Point Summation.
+ *      2005-11-13, Tech. Rep. 05.1, Fac. Inf. Commun. Sci, Hambg. Univ.
+ *      Technol., also known as SIAM J. Sci. Comput. Volume 31, Issue 1, pp.
+ *      189-224 (2008),
+ *      http://www.ti3.tuhh.de/paper/rump/RuOgOi07I.pdf
+ *
+ * The main purpose of these templates is to implement Algorithm 6.1 (here as
+ * acc_sum) which is required by compensated complex floating-point polynomial
+ * evaluation.
+ */
+namespace ellint_carlson { namespace arithmetic { namespace aux
+{
+    /* Algorithm 3.6 */
+    template<typename RT>
+    inline typing::real_only<RT, RT>
+    next_power_two(const RT& p)
+    {
+	RT q = p / typing::half_epsilon<RT>();
+	RT L = std::abs((q + p) - q);
+	if ( L == 0.0 )
+	{
+	    L = std::abs(p);
+	}
+	return L;
+    }
+
+
+    /* Algorithm 3.2 */
+    template<typename RT>
+    inline typing::real_only<RT, void>
+    extract_scalar(const RT& sigma, RT& p, RT& q)
+    {
+	q = (sigma + p) - sigma;
+	p -= q;
+    }
+
+
+    /* Algorithm 3.4 */
+    template<typename RT, std::size_t LEN>
+    inline typing::real_only<RT, void>
+    extract_vector(const RT& sigma, RT(& p)[LEN], bool(& mask)[LEN], RT& tau)
+    {
+	RT q;
+	tau = (RT)0.0;
+	for ( std::size_t i = 0; i < LEN; ++i )
+	{
+	    if ( mask[i] )
```

Comment:
I don't know if you're worried about speed (and you'd definitely have to benchmark solutions like the one I'm about to tell you), but there a lot of instances in this code where you can remove branch points (especially in loops) and replace them with something like this:
```cpp
for ( std::size_t i = 0; i < LEN; ++i )
{
    extract_scalar(sigma, p[i], q);
    tau += mask[i]*q;
    mask[i] &= p[i] != 0;
}
```
Again, only if speed is an issue and bench-marking is showing significant improvement doing it this way as opposed to nested if statements.